graph/graph.go

```diff
@@ -27,11 +28,64 @@ import (
 	"github.com/docker/docker/runconfig"
 )
 
+type retainedLayers struct {
+	layers map[string][]string
+	sync.Mutex
+}
+
+func (r *retainedLayers) Add(layerIDs []string, imgID string) {
+	r.Lock()
+	defer r.Unlock()
+	for _, layerID := range layerIDs {
+		if imgs, exists := r.layers[layerID]; exists {
+			r.layers[layerID] = append(imgs, imgID)
+		} else {
+			r.layers[layerID] = []string{imgID}
+		}
+	}
+}
+
+func (r *retainedLayers) Delete(layerIDs []string, imgID string) {
```

Comment:
I propose to change signature to `(imgID string, layers ...string)`. Then you don't need to write stuff like `[]string{..}`.